src/PyCall.jl

```diff
@@ -438,6 +438,73 @@ end
 #########################################################################
 
 """
+    @pywith
+
+Mimics a Python 'with' statement. Usage:
+
+@pywith EXPR[::TYPE1] [as VAR[::TYPE2]] begin
+    BLOCK
+end
+
+TYPE1/TYPE2 can optionally be used to override automatic conversion to Julia
+types for both the context manager and variable in cases where this is not
+desired.
+
+"""
+macro pywith(EXPR,as,VAR,BLOCK)
+    if isexpr(VAR,:(::))
+        VAR,TYPE = VAR.args
+    else
+        TYPE = PyAny
+    end
+    if !((as==:as) && isa(VAR,Symbol))
+        throw(ArgumentError("usage: @pywith EXPR[::TYPE1] [as VAR[::TYPE2]] BLOCK."))
+    end
+    _pywith(EXPR,VAR,TYPE,BLOCK)
+end
+macro pywith(EXPR,BLOCK)
+    _pywith(EXPR,nothing,PyObject,BLOCK)
+end
+
+
+function _pywith(EXPR,VAR,TYPE,BLOCK)
+    EXPR_str = string(EXPR)
+    quote
+        mgr = $(isexpr(EXPR,:(::)) ? :(@pycall $(esc(EXPR))) : esc(EXPR))
+        if !isa(mgr,PyObject)
+            if $(isexpr(EXPR,:call))
+                throw(ArgumentError("@pywith: `$($EXPR_str)` did not return a PyObject. If this is a call to a Python function, try `$($EXPR_str)::PyObject` to turn off automatic conversion."))
+            else
+                throw(ArgumentError("@pywith: `$($EXPR_str)` should be a PyObject."))
+            end
+        end
+        mgrT = pytypeof(mgr)
+        exit = mgrT["__exit__"]
+        value = @pycall mgrT["__enter__"](mgr)::$(esc(TYPE))
+        exc = true
+        try
+            try
+                $(VAR==nothing ? :() : :($(esc(VAR)) = value))
+                $(esc(BLOCK))
+            catch err
+                exc = false
+                if !(@pycall exit(mgr, pyimport(:sys)[:exc_info]()...)::Bool)
+                    throw(err)
+                end
+            end
+        finally
+            if exc
+                exit(mgr, nothing, nothing, nothing)
+            end
+        end
+        nothing
```

Comment:
The reasoning here btw was that Python with-statements don't return a value, so this is for consistency. You could certainly let them return something in Julia, but then you have to make a decision what to return in case there was an error that was swallowed up by the context manager's exit function. I figured this was just simpler.